File: ScriptProject/Missing/System/BitConverter.cs
﻿namespace System
{
    using System.Runtime.CompilerServices;

    [Imported]
    [ModuleName(null)]
    [IgnoreNamespace]
    public static class BitConverter
    {
        public static float ToSingle(Uint8Array value, int startIndex)
        {
            throw new ImplementedInJavaScript();
        }

        public static int ToInt32(Uint8Array value, int startIndex)
        {
            throw new ImplementedInJavaScript();
        }

        public static uint ToUInt32(Uint8Array value, int startIndex)
        {
            throw new ImplementedInJavaScript();
        }

        public static ushort ToUInt16(Uint8Array value, int startIndex)
        {
            throw new ImplementedInJavaScript();
        }

        public static short ToInt16(Uint8Array value, int startIndex)
        {
            throw new ImplementedInJavaScript();
        }

        [ScriptName("getBytesFromFloat")]
        public static Uint8Array GetBytes(float value)
        {
            throw new ImplementedInJavaScript();
        }

        [ScriptName("getBytesFromInt")]
        public static Uint8Array GetBytes(int value)
        {
            throw new ImplementedInJavaScript();
        }

        [ScriptName("getBytesFromInt")]
        public static Uint8Array GetBytes(uint value)
        {
            throw new ImplementedInJavaScript();
        }
    }
}

File: web/scripts/missing/BitConverter.js
BitConverter = {
    toUInt32: function(/*byte[]*/ value, /*int*/ startIndex) {
        var ds = new DataStream(value, true);
        ds.position = startIndex;
        var val = ds.readUint32();
        return val;
    },
    toInt32: function(/*byte[]*/ value, /*int*/ startIndex) {
        var ds = new DataStream(value, true);
        ds.position = startIndex;
        var val = ds.readInt32();
        return val;
    },
    toUInt16: function(/*byte[]*/ value, /*int*/ startIndex) {
        var ds = new DataStream(value, true);
        ds.position = startIndex;
        var val = ds.readUint16();
        return val;
    },
    toInt16: function(/*byte[]*/ value, /*int*/ startIndex) {
        var ds = new DataStream(value, true);
        ds.position = startIndex;
        var val = ds.readInt16();
        return val;
    },
    toSingle: function(/*byte[]*/ value, /*int*/ startIndex) {
        var ds = new DataStream(value, true);
        ds.position = startIndex;
        var val = ds.readFloat32();
        return val;
    },
    getBytesFromFloat: function(f) {
        var view = new DataView(new ArrayBuffer(4));
        view.setFloat32(0, f, true);
        return new Uint8Array(view.buffer);
    },
    getBytesFromInt: function(i) {
        var view = new DataView(new ArrayBuffer(4));
        view.setUint32(0, i, true);
        return new Uint8Array(view.buffer);
    }
};
